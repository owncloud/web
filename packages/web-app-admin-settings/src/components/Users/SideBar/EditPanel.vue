<template>
  <div id="user-edit-panel" class="oc-mt-xl">
    <UserInfoBox :user="user" />
    <form id="user-edit-form" class="oc-background-highlight oc-p-m" autocomplete="off">
      <div>
        <oc-text-input
          id="userName-input"
          v-model="editUser.onPremisesSamAccountName"
          class="oc-mb-s"
          :label="$gettext('User name')"
          :error-message="formData.userName.errorMessage"
          :fix-message-line="true"
          :read-only="isInputFieldReadOnly('user.onPremisesSamAccountName')"
          @update:model-value="validateUserName"
        />
        <oc-text-input
          id="displayName-input"
          v-model="editUser.displayName"
          class="oc-mb-s"
          :label="$gettext('First and last name')"
          :error-message="formData.displayName.errorMessage"
          :fix-message-line="true"
          :read-only="isInputFieldReadOnly('user.displayName')"
          @update:model-value="validateDisplayName"
        />
        <oc-text-input
          id="email-input"
          v-model="editUser.mail"
          class="oc-mb-s"
          :label="$gettext('Email')"
          :error-message="formData.email.errorMessage"
          type="email"
          :fix-message-line="true"
          :read-only="isInputFieldReadOnly('user.mail')"
          @change="validateEmail"
        />
        <oc-text-input
          id="password-input"
          :model-value="editUser.passwordProfile?.password"
          class="oc-mb-s"
          :label="$gettext('Password')"
          type="password"
          :fix-message-line="true"
          placeholder="●●●●●●●●"
          :read-only="isInputFieldReadOnly('user.passwordProfile')"
          @update:model-value="onUpdatePassword"
        />
        <div class="oc-mb-s">
          <oc-select
            id="role-input"
            :model-value="selectedRoleValue"
            :label="$gettext('Role')"
            option-label="displayName"
            :options="translatedRoleOptions"
            :clearable="false"
            :read-only="isInputFieldReadOnly('user.appRoleAssignments')"
            @update:model-value="onUpdateRole"
          />
          <div class="oc-text-input-message"></div>
        </div>
        <div class="oc-mb-s">
          <oc-select
            id="login-input"
            :disabled="isLoginInputDisabled"
            :model-value="selectedLoginValue"
            :label="$gettext('Login')"
            :options="loginOptions"
            :clearable="false"
            :read-only="isInputFieldReadOnly('user.accountEnabled')"
            @update:model-value="onUpdateLogin"
          />

          <div class="oc-text-input-message"></div>
        </div>
        <quota-select
          id="quota-select-form"
          :key="'quota-select-' + user.id"
          :disabled="isQuotaInputDisabled"
          class="oc-mb-s"
          :label="$gettext('Personal quota')"
          :total-quota="editUser.drive?.quota?.total || 0"
          :max-quota="maxQuota"
          :fix-message-line="true"
          :description-message="
            isQuotaInputDisabled && !isInputFieldReadOnly('drive.quota')
              ? $gettext('To set an individual quota, the user needs to have logged in once.')
              : ''
          "
          :read-only="isInputFieldReadOnly('drive.quota')"
          @selected-option-change="changeSelectedQuotaOption"
        />
        <group-select
          class="oc-mb-s"
          :read-only="isInputFieldReadOnly('user.memberOf')"
          :selected-groups="editUser.memberOf"
          :group-options="groupOptions"
          @selected-option-change="changeSelectedGroupOption"
        />
      </div>
      <compare-save-dialog
        class="edit-compare-save-dialog oc-mb-l"
        :original-object="compareSaveDialogOriginalObject"
        :compare-object="editUser"
        :confirm-button-disabled="invalidFormData"
        @revert="revertChanges"
        @confirm="onEditUser({ user, editUser })"
      ></compare-save-dialog>
    </form>
  </div>
</template>
<script lang="ts">
import { computed, defineComponent, PropType, ref, unref } from 'vue'
import * as EmailValidator from 'email-validator'
import UserInfoBox from './UserInfoBox.vue'
import {
  CompareSaveDialog,
  QuotaSelect,
  useUserStore,
  useCapabilityStore,
  useEventBus,
  useMessages,
  useSpacesStore,
  useSharesStore
} from '@ownclouders/web-pkg'
import GroupSelect from '../GroupSelect.vue'
import { cloneDeep, isEmpty, isEqual, omit } from 'lodash-es'
import { AppRole, AppRoleAssignment, Group, User } from '@ownclouders/web-client/graph/generated'
import { MaybeRef, useClientService } from '@ownclouders/web-pkg'
import { storeToRefs } from 'pinia'
import { diff } from 'deep-object-diff'
import { useUserSettingsStore } from '../../../composables/stores/userSettings'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'EditPanel',
  components: {
    UserInfoBox,
    CompareSaveDialog,
    QuotaSelect,
    GroupSelect
  },
  props: {
    user: {
      type: Object as PropType<User>,
      required: false,
      default: null
    },
    roles: {
      type: Array as PropType<AppRole[]>,
      required: true
    },
    groups: {
      type: Array as PropType<Group[]>,
      required: true
    },
    applicationId: {
      type: String,
      required: true
    }
  },
  emits: ['confirm'],
  setup(props) {
    const capabilityStore = useCapabilityStore()
    const capabilityRefs = storeToRefs(capabilityStore)
    const clientService = useClientService()
    const userStore = useUserStore()
    const userSettingsStore = useUserSettingsStore()
    const spacesStore = useSpacesStore()
    const sharesStore = useSharesStore()
    const eventBus = useEventBus()
    const { showErrorMessage } = useMessages()
    const { $gettext } = useGettext()

    const editUser: MaybeRef<User> = ref()
    const formData = ref({
      displayName: {
        errorMessage: '',
        valid: true
      },
      userName: {
        errorMessage: '',
        valid: true
      },
      email: {
        errorMessage: '',
        valid: true
      }
    })
    const groupOptions = computed(() => {
      const { memberOf: selectedGroups } = unref(editUser)
      return props.groups.filter(
        (g) => !selectedGroups.some((s) => s.id === g.id) && !g.groupTypes?.includes('ReadOnly')
      )
    })
    const isLoginInputDisabled = computed(() => userStore.user.id === (props.user as User).id)
    const isInputFieldReadOnly = (key: string) => {
      return capabilityStore.graphUsersReadOnlyAttributes.includes(key)
    }

    const onUpdateUserAppRoleAssignments = (user: User, editUser: User) => {
      const client = clientService.graphAuthenticated
      return client.users.createUserAppRoleAssignment(user.id, {
        appRoleId: editUser.appRoleAssignments[0].appRoleId,
        resourceId: props.applicationId,
        principalId: editUser.id
      })
    }
    const onUpdateUserGroupAssignments = (user: User, editUser: User) => {
      const client = clientService.graphAuthenticated
      const groupsToAdd = editUser.memberOf.filter(
        (editUserGroup) => !user.memberOf.some((g) => g.id === editUserGroup.id)
      )
      const groupsToDelete = user.memberOf.filter(
        (editUserGroup) => !editUser.memberOf.some((g) => g.id === editUserGroup.id)
      )
      const requests = []

      for (const groupToAdd of groupsToAdd) {
        requests.push(client.groups.addMember(groupToAdd.id, user.id))
      }
      for (const groupToDelete of groupsToDelete) {
        requests.push(client.groups.deleteMember(groupToDelete.id, user.id))
      }

      return Promise.all(requests)
    }

    const onUpdateUserDrive = async (editUser: User) => {
      const client = clientService.graphAuthenticated
      const updateSpace = await client.drives.updateDrive(
        editUser.drive.id,
        {
          name: editUser.drive.name,
          quota: { total: editUser.drive.quota.total }
        },
        sharesStore.graphRoles
      )

      if (editUser.id === userStore.user.id) {
        // Load current user quota
        spacesStore.updateSpaceField({
          id: editUser.drive.id,
          field: 'spaceQuota',
          value: updateSpace.spaceQuota
        })
      }
    }

    const onEditUser = async ({ user, editUser }: { user: User; editUser: User }) => {
      try {
        const client = clientService.graphAuthenticated
        const graphEditUserPayloadExtractor = (user: User) => {
          return omit(user, ['drive', 'appRoleAssignments', 'memberOf'])
        }
        const graphEditUserPayload = diff(
          graphEditUserPayloadExtractor(user),
          graphEditUserPayloadExtractor(editUser)
        ) as User

        if (!isEmpty(graphEditUserPayload)) {
          await client.users.editUser(editUser.id, graphEditUserPayload)
        }

        if (!isEqual(user.drive?.quota?.total, editUser.drive?.quota?.total)) {
          await onUpdateUserDrive(editUser)
        }

        if (!isEqual(user.memberOf, editUser.memberOf)) {
          await onUpdateUserGroupAssignments(user, editUser)
        }

        if (
          !isEqual(user.appRoleAssignments[0]?.appRoleId, editUser.appRoleAssignments[0]?.appRoleId)
        ) {
          await onUpdateUserAppRoleAssignments(user, editUser)
        }

        const updatedUser = await client.users.getUser(user.id)
        userSettingsStore.upsertUser(updatedUser)

        eventBus.publish('sidebar.entity.saved')

        if (userStore.user.id === updatedUser.id) {
          userStore.setUser(updatedUser)
        }

        return updatedUser
      } catch (error) {
        console.error(error)
        showErrorMessage({
          title: $gettext('Failed to edit user'),
          errors: [error]
        })
      }
    }

    return {
      maxQuota: capabilityRefs.spacesMaxQuota,
      isInputFieldReadOnly,
      isLoginInputDisabled,
      editUser,
      formData,
      groupOptions,
      clientService,
      onEditUser,
      // HACK: make sure _user has a proper type
      _user: computed(() => props.user as User)
    }
  },
  computed: {
    loginOptions() {
      return [
        {
          label: this.$gettext('Allowed'),
          value: true
        },
        {
          label: this.$gettext('Forbidden'),
          value: false
        }
      ]
    },
    selectedLoginValue() {
      return this.loginOptions.find((option) =>
        !('accountEnabled' in this.editUser)
          ? option.value === true
          : this.editUser.accountEnabled === option.value
      )
    },
    translatedRoleOptions() {
      return this.roles.map((role) => {
        return { ...role, displayName: this.$gettext(role.displayName) }
      })
    },
    selectedRoleValue() {
      const assignedRole = this.editUser?.appRoleAssignments?.[0]
      return this.translatedRoleOptions.find((role) => role.id === assignedRole?.appRoleId)
    },
    invalidFormData() {
      return Object.values(this.formData)
        .map((v: any) => !!v.valid)
        .includes(false)
    },
    showQuota() {
      return this.editUser.drive?.quota
    },
    isQuotaInputDisabled() {
      return typeof this.showQuota === 'undefined'
    },
    compareSaveDialogOriginalObject() {
      return cloneDeep(this.user)
    }
  },
  watch: {
    user: {
      handler: function () {
        this.editUser = cloneDeep(this.user)
      },
      deep: true,
      immediate: true
    },
    editUser: {
      handler: function () {
        /**
         * Property accountEnabled won't be always set, but this still means, that login is allowed.
         * So we actually don't need to change the property if missing and not set to forbidden in the UI.
         * This also avoids the compare save dialog from displaying that there are unsaved changes.
         */
        if (this.editUser.accountEnabled === true && !('accountEnabled' in this.user)) {
          delete this.editUser.accountEnabled
        }
      },
      deep: true
    }
  },
  methods: {
    changeSelectedQuotaOption(option: { value: number; displayValue: string }) {
      this.editUser.drive.quota.total = option.value
    },
    changeSelectedGroupOption(option: Group[]) {
      this.editUser.memberOf = option
    },
    async validateUserName() {
      this.formData.userName.valid = false

      if (this.editUser.onPremisesSamAccountName.trim() === '') {
        this.formData.userName.errorMessage = this.$gettext('User name cannot be empty')
        return false
      }

      if (this.editUser.onPremisesSamAccountName.includes(' ')) {
        this.formData.userName.errorMessage = this.$gettext('User name cannot contain white spaces')
        return false
      }

      if (
        this.editUser.onPremisesSamAccountName.length &&
        !isNaN(parseInt(this.editUser.onPremisesSamAccountName[0]))
      ) {
        this.formData.userName.errorMessage = this.$gettext('User name cannot start with a number')
        return false
      }

      if (this.editUser.onPremisesSamAccountName.length > 255) {
        this.formData.userName.errorMessage = this.$gettext(
          'User name cannot exceed 255 characters'
        )
        return false
      }

      if (this.user.onPremisesSamAccountName !== this.editUser.onPremisesSamAccountName) {
        try {
          // Validate username by fetching the user. If the request succeeds, we throw a validation error
          const client = this.clientService.graphAuthenticated
          await client.users.getUser(this.editUser.onPremisesSamAccountName)
          this.formData.userName.errorMessage = this.$gettext('User "%{userName}" already exists', {
            userName: this.editUser.onPremisesSamAccountName
          })
          return false
        } catch {}
      }

      this.formData.userName.errorMessage = ''
      this.formData.userName.valid = true
      return true
    },
    validateDisplayName() {
      this.formData.displayName.valid = false

      if (this.editUser.displayName.trim() === '') {
        this.formData.displayName.errorMessage = this.$gettext(
          'First and last name cannot be empty'
        )
        return false
      }

      if (this.editUser.displayName.length > 255) {
        this.formData.displayName.errorMessage = this.$gettext(
          'First and last name cannot exceed 255 characters'
        )
        return false
      }

      this.formData.displayName.errorMessage = ''
      this.formData.displayName.valid = true
      return true
    },
    validateEmail() {
      this.formData.email.valid = false

      if (!EmailValidator.validate(this.editUser.mail)) {
        this.formData.email.errorMessage = this.$gettext('Please enter a valid email')
        return false
      }

      this.formData.email.errorMessage = ''
      this.formData.email.valid = true
      return true
    },
    revertChanges() {
      this.editUser = cloneDeep(this.user)
      Object.values(this.formData).forEach((formDataValue: any) => {
        formDataValue.valid = true
        formDataValue.errorMessage = ''
      })
    },
    onUpdateRole(role: AppRoleAssignment) {
      if (!this.editUser.appRoleAssignments.length) {
        // FIXME: Add resourceId and principalId to be able to remove type cast
        this.editUser.appRoleAssignments.push({
          appRoleId: role.id
        } as AppRoleAssignment)
        return
      }
      this.editUser.appRoleAssignments[0].appRoleId = role.id
    },
    onUpdatePassword(password: string) {
      this.editUser.passwordProfile = {
        password
      }
    },
    onUpdateLogin({ value }: { value: boolean }) {
      this.editUser.accountEnabled = value
    }
  }
})
</script>

<style lang="scss">
#user-edit-panel {
  #user-edit-form {
    border-radius: 5px;
  }
}
</style>
