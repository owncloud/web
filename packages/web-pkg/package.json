{
  "name": "@ownclouders/web-pkg",
  "version": "11.1.3",
  "description": "ownCloud web pkg",
  "license": "AGPL-3.0",
  "main": "src/index.ts",
  "types": "types.d.ts",
  "private": false,
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "type": "module",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-pkg",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-pkg"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "import": "./dist/web-pkg.js",
        "require": "./dist/web-pkg.umd.cjs",
        "types": "./dist/src/index.d.ts"
      }
    }
  },
  "scripts": {
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && rm -rf package/dist/tests package/dist/*.css && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@casl/vue": "^2.2.2",
    "@microsoft/fetch-event-source": "^2.0.1",
    "@ownclouders/design-system": "workspace:^",
    "@ownclouders/web-client": "workspace:^",
    "@sentry/vue": "^8.33.1",
    "@uppy/core": "^3.3.0",
    "@uppy/drop-target": "^2.0.0",
    "@uppy/tus": "^3.1.0",
    "@uppy/utils": "^5.3.0",
    "@uppy/xhr-upload": "^3.0.1",
    "@vavt/cm-extension": "^1.6.0",
    "@vue/shared": "^3.5.11",
    "@vueuse/core": "^11.0.0",
    "axios": "^1.7.7",
    "deepmerge": "^4.2.2",
    "dompurify": "^3.2.3",
    "filesize": "^10.1.0",
    "fuse.js": "^7.0.0",
    "js-generate-password": "^1.0.0",
    "lodash-es": "^4.17.21",
    "luxon": "^3.5.0",
    "mark.js": "^8.11.1",
    "md-editor-v3": "^5.2.1",
    "oidc-client-ts": "^2.4.0 || ^3.0.0",
    "p-queue": "^8.0.0",
    "password-sheriff": "^1.1.1",
    "pinia": "^2.2.4",
    "portal-vue": "^3.0.0",
    "prismjs": "^1.29.0",
    "qs": "^6.13.0",
    "semver": "^7.6.3",
    "uuid": "^11.0.0",
    "vue-concurrency": "^5.0.1",
    "vue-router": "^4.2.5",
    "vue3-gettext": "^2.4.0",
    "zod": "^3.23.8"
  },
  "devDependencies": {
    "@ownclouders/web-test-helpers": "workspace:^",
    "@types/dompurify": "3.0.5",
    "@types/lodash-es": "4.17.12",
    "@vitest/web-worker": "2.1.2",
    "clean-publish": "5.0.0",
    "vite-plugin-dts": "4.2.4",
    "vite-plugin-node-polyfills": "0.22.0"
  }
}
