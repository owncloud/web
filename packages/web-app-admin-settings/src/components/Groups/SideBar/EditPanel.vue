<template>
  <div id="group-edit-panel" class="oc-mt-xl">
    <group-info-box :group="group" />
    <form id="group-edit-form" class="oc-background-highlight oc-p-m" autocomplete="off">
      <oc-text-input
        id="displayName-input"
        v-model="editGroup.displayName"
        class="oc-mb-s"
        :label="$gettext('Group name')"
        :error-message="formData.displayName.errorMessage"
        :fix-message-line="true"
        @update:model-value="validateDisplayName"
      />
      <compare-save-dialog
        class="edit-compare-save-dialog oc-mb-l"
        :original-object="group"
        :compare-object="editGroup"
        :confirm-button-disabled="invalidFormData"
        @revert="revertChanges"
        @confirm="onEditGroup(editGroup)"
      ></compare-save-dialog>
    </form>
  </div>
</template>
<script lang="ts">
import { defineComponent, PropType, ref } from 'vue'
import { Group } from '@ownclouders/web-client/graph/generated'
import { CompareSaveDialog, eventBus, useMessages } from '@ownclouders/web-pkg'
import { MaybeRef, useClientService } from '@ownclouders/web-pkg'
import GroupInfoBox from './GroupInfoBox.vue'
import { useGroupSettingsStore } from '../../../composables'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'EditPanel',
  components: {
    GroupInfoBox,
    CompareSaveDialog
  },
  props: {
    group: {
      type: Object as PropType<Group>,
      required: true,
      default: null
    }
  },
  emits: ['confirm'],
  setup() {
    const clientService = useClientService()
    const { showErrorMessage } = useMessages()
    const groupSettingsStore = useGroupSettingsStore()
    const { $gettext } = useGettext()

    const editGroup: MaybeRef<Group> = ref({})
    const formData = ref({
      displayName: {
        errorMessage: '',
        valid: true
      }
    })

    const onEditGroup = async (editGroup: Group) => {
      try {
        const client = clientService.graphAuthenticated
        await client.groups.editGroup(editGroup.id, editGroup)
        const updatedGroup = await client.groups.getGroup(editGroup.id)
        groupSettingsStore.upsertGroup(updatedGroup)

        eventBus.publish('sidebar.entity.saved')

        return updatedGroup
      } catch (error) {
        console.error(error)
        showErrorMessage({
          title: $gettext('Failed to edit group'),
          errors: [error]
        })
      }
    }

    return {
      clientService,
      editGroup,
      formData,
      onEditGroup
    }
  },
  computed: {
    invalidFormData() {
      return Object.values(this.formData)
        .map((v: any) => !!v.valid)
        .includes(false)
    }
  },
  watch: {
    group: {
      handler: function () {
        this.editGroup = { ...this.group }
      },
      deep: true,
      immediate: true
    }
  },
  methods: {
    async validateDisplayName() {
      this.formData.displayName.valid = false

      if (this.editGroup.displayName.trim() === '') {
        this.formData.displayName.errorMessage = this.$gettext('Group name cannot be empty')
        return false
      }

      if (this.editGroup.displayName.length > 255) {
        this.formData.displayName.errorMessage = this.$gettext(
          'Group name cannot exceed 255 characters'
        )
        return false
      }

      if (this.group.displayName !== this.editGroup.displayName) {
        try {
          const client = this.clientService.graphAuthenticated
          await client.groups.getGroup(this.editGroup.displayName)
          this.formData.displayName.errorMessage = this.$gettext(
            'Group "%{groupName}" already exists',
            {
              groupName: this.editGroup.displayName
            }
          )
          return false
        } catch {}
      }

      this.formData.displayName.errorMessage = ''
      this.formData.displayName.valid = true
      return true
    },
    revertChanges() {
      this.editGroup = { ...this.group }
      Object.values(this.formData).forEach((formDataValue: any) => {
        formDataValue.valid = true
        formDataValue.errorMessage = ''
      })
    }
  }
})
</script>
<style lang="scss">
#group-edit-panel {
  #group-edit-form {
    border-top-left-radius: 5px;
    border-top-right-radius: 5px;
  }

  .edit-compare-save-dialog {
    border-bottom-left-radius: 5px;
    border-bottom-right-radius: 5px;
  }

  .group-info {
    align-items: center;
    flex-direction: column;
  }

  .group-info-display-name {
    font-size: 1.5rem;
  }
}
</style>
