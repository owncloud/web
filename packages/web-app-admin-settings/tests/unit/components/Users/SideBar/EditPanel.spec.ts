import EditPanel from '../../../../../src/components/Users/SideBar/EditPanel.vue'
import {
  defaultComponentMocks,
  defaultPlugins,
  mockAxiosReject,
  shallowMount
} from '@ownclouders/web-test-helpers'
import { mock } from 'vitest-mock-extended'
import { Drive, Group, User } from '@ownclouders/web-client/graph/generated'
import { CapabilityStore } from '@ownclouders/web-pkg'
import GroupSelect from '../../../../../src/components/Users/GroupSelect.vue'

const availableGroupOptions = [
  mock<Group>({ id: '1', displayName: 'group1', groupTypes: [] }),
  mock<Group>({ id: '2', displayName: 'group2', groupTypes: [] })
]
const selectors = {
  groupSelectStub: 'group-select-stub'
}

describe('EditPanel', () => {
  it('renders all available inputs', () => {
    const { wrapper } = getWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('filters selected groups when passing the options to the GroupSelect component', () => {
    const { wrapper } = getWrapper({ selectedGroups: [availableGroupOptions[0]] })
    const selectedGroups = wrapper
      .findComponent<typeof GroupSelect>(selectors.groupSelectStub)
      .props('selectedGroups')
    const groupOptions = wrapper
      .findComponent<typeof GroupSelect>(selectors.groupSelectStub)
      .props('groupOptions')
    expect(selectedGroups.length).toBe(1)
    expect(selectedGroups[0].id).toEqual(availableGroupOptions[0].id)
    expect(groupOptions.length).toBe(1)
    expect(groupOptions[0].id).toEqual(availableGroupOptions[1].id)
  })

  describe('method "isInputFieldReadOnly"', () => {
    it('should be true if included in capability readOnlyUserAttributes list', () => {
      const { wrapper } = getWrapper({ readOnlyUserAttributes: ['user.displayName'] })
      expect(wrapper.vm.isInputFieldReadOnly('user.displayName')).toBeTruthy()
    })
    it('should be false if not included in capability readOnlyUserAttributes list', () => {
      const { wrapper } = getWrapper()
      expect(wrapper.vm.isInputFieldReadOnly('user.displayName')).toBeFalsy()
    })
  })

  describe('method "revertChanges"', () => {
    it('should revert changes on property editUser', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = 'jana'
      wrapper.vm.editUser.mail = 'jana@owncloud.com'
      wrapper.vm.revertChanges()
      expect(wrapper.vm.editUser.displayName).toEqual('jan')
      expect(wrapper.vm.editUser.mail).toEqual('jan@owncloud.com')
    })
    it('should revert changes on property formData', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = false
      wrapper.vm.formData.displayName.errorMessage = 'error'
      wrapper.vm.revertChanges()
      expect(wrapper.vm.formData.displayName.valid).toBeTruthy()
      expect(wrapper.vm.formData.displayName.errorMessage).toEqual('')
    })
  })

  describe('method "validateUserName"', () => {
    it('should be false when userName is empty', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.onPremisesSamAccountName = ''
      expect(await wrapper.vm.validateUserName()).toBeFalsy()
    })
    it('should be false if userName is longer than 255 characters', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.onPremisesSamAccountName = 'n'.repeat(256)
      expect(await wrapper.vm.validateUserName()).toBeFalsy()
    })
    it('should be false when userName contains white spaces', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.onPremisesSamAccountName = 'jan owncCloud'
      expect(await wrapper.vm.validateUserName()).toBeFalsy()
    })
    it('should be false when userName starts with a numeric value', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.onPremisesSamAccountName = '1moretry'
      expect(await wrapper.vm.validateUserName()).toBeFalsy()
    })
    it('should be false when userName is already existing', async () => {
      const { wrapper, mocks } = getWrapper()
      const graphMock = mocks.$clientService.graphAuthenticated
      const getUserStub = graphMock.users.getUser.mockResolvedValue(
        mock<User>({ onPremisesSamAccountName: 'jan' })
      )
      wrapper.vm.editUser.onPremisesSamAccountName = 'jan'
      expect(await wrapper.vm.validateUserName()).toBeFalsy()
      expect(getUserStub).toHaveBeenCalled()
    })
    it('should be true when userName is valid', async () => {
      const { wrapper, mocks } = getWrapper()
      const graphMock = mocks.$clientService.graphAuthenticated
      const getUserStub = graphMock.users.getUser.mockRejectedValue(() => mockAxiosReject())
      wrapper.vm.editUser.onPremisesSamAccountName = 'jana'
      expect(await wrapper.vm.validateUserName()).toBeTruthy()
      expect(getUserStub).toHaveBeenCalled()
    })
  })

  describe('method "validateDisplayName"', () => {
    it('should return true if displayName is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = 'jan'
      expect(wrapper.vm.validateDisplayName()).toBeTruthy()
      expect(wrapper.vm.formData.displayName.valid).toBeTruthy()
    })
    it('should be false if displayName is longer than 255 characters', async () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = 'n'.repeat(256)
      expect(await wrapper.vm.validateDisplayName()).toBeFalsy()
    })
    it('should return false if displayName is not valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.displayName = ''
      expect(wrapper.vm.validateDisplayName()).toBeFalsy()
      expect(wrapper.vm.formData.displayName.valid).toBeFalsy()
    })
  })

  describe('method "validateEmail"', () => {
    it('should return true if email is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.mail = 'jan@owncloud.com'
      expect(wrapper.vm.validateEmail()).toBeTruthy()
      expect(wrapper.vm.formData.email.valid).toBeTruthy()
    })
    it('should return false if email is not valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.editUser.mail = ''
      expect(wrapper.vm.validateEmail()).toBeFalsy()
      expect(wrapper.vm.formData.email.valid).toBeFalsy()
    })
  })

  describe('computed method "invalidFormData"', () => {
    it('should be false if formData is invalid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = true
      expect(wrapper.vm.invalidFormData).toBeFalsy()
    })
    it('should be true if formData is valid', () => {
      const { wrapper } = getWrapper()
      wrapper.vm.formData.displayName.valid = false
      expect(wrapper.vm.invalidFormData).toBeTruthy()
    })
  })

  describe('group select', () => {
    it('takes all available groups', () => {
      const { wrapper } = getWrapper()
      expect(
        wrapper.findComponent<typeof GroupSelect>('group-select-stub').props('groupOptions').length
      ).toBe(availableGroupOptions.length)
    })
    it('filters out read-only groups', () => {
      const { wrapper } = getWrapper({
        groups: [mock<Group>({ id: '1', displayName: 'group1', groupTypes: ['ReadOnly'] })]
      })
      expect(
        wrapper.findComponent<typeof GroupSelect>('group-select-stub').props('groupOptions').length
      ).toBe(0)
    })
  })
})

function getWrapper({
  readOnlyUserAttributes = [],
  selectedGroups = [],
  groups = availableGroupOptions
}: { readOnlyUserAttributes?: string[]; selectedGroups?: Group[]; groups?: Group[] } = {}) {
  const mocks = defaultComponentMocks()
  const capabilities = {
    graph: { users: { read_only_attributes: readOnlyUserAttributes }, tags: { max_tag_length: 30 } }
  } satisfies Partial<CapabilityStore['capabilities']>

  return {
    mocks,
    wrapper: shallowMount(EditPanel, {
      props: {
        user: {
          id: '2',
          displayName: 'jan',
          mail: 'jan@owncloud.com',
          passwordProfile: { password: '' },
          drive: { quota: {} } as Drive,
          memberOf: selectedGroups
        } as User,
        roles: [{ id: '1', displayName: 'admin' }],
        groups,
        applicationId: '1'
      },
      global: {
        mocks,
        provide: mocks,
        plugins: [...defaultPlugins({ piniaOptions: { capabilityState: { capabilities } } })]
      }
    })
  }
}
