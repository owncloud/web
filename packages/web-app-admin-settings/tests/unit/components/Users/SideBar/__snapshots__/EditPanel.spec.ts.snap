// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditPanel > renders all available inputs 1`] = `
"<div id="user-edit-panel" class="oc-mt-xl">
  <user-info-box-stub user="[object Object]"></user-info-box-stub>
  <form id="user-edit-form" class="oc-background-highlight oc-p-m" autocomplete="off">
    <div>
      <oc-text-input-stub id="userName-input" type="text" clearbuttonenabled="false" clearbuttonaccessiblelabel="" disabled="false" label="User name" errormessage="" fixmessageline="true" readonly="false" passwordpolicy="[object Object]" class="oc-mb-s"></oc-text-input-stub>
      <oc-text-input-stub id="displayName-input" type="text" modelvalue="jan" clearbuttonenabled="false" clearbuttonaccessiblelabel="" disabled="false" label="First and last name" errormessage="" fixmessageline="true" readonly="false" passwordpolicy="[object Object]" class="oc-mb-s"></oc-text-input-stub>
      <oc-text-input-stub id="email-input" type="email" modelvalue="jan@owncloud.com" clearbuttonenabled="false" clearbuttonaccessiblelabel="" disabled="false" label="Email" errormessage="" fixmessageline="true" readonly="false" passwordpolicy="[object Object]" class="oc-mb-s"></oc-text-input-stub>
      <oc-text-input-stub id="password-input" type="password" modelvalue="" clearbuttonenabled="false" clearbuttonaccessiblelabel="" disabled="false" label="Password" fixmessageline="true" readonly="false" passwordpolicy="[object Object]" class="oc-mb-s" placeholder="●●●●●●●●"></oc-text-input-stub>
      <div class="oc-mb-s">
        <oc-select-stub id="role-input" filter="[Function]" disabled="false" label="Role" optionlabel="displayName" getoptionlabel="[Function]" searchable="true" clearable="false" loading="false" fixmessageline="false" multiple="false" readonly="false" positionfixed="false" options="[object Object]"></oc-select-stub>
        <div class="oc-text-input-message"></div>
      </div>
      <div class="oc-mb-s">
        <oc-select-stub id="login-input" filter="[Function]" disabled="false" label="Login" optionlabel="label" getoptionlabel="[Function]" searchable="true" clearable="false" loading="false" fixmessageline="false" multiple="false" readonly="false" positionfixed="false" model-value="[object Object]" options="[object Object],[object Object]"></oc-select-stub>
        <div class="oc-text-input-message"></div>
      </div>
      <quota-select-stub totalquota="0" maxquota="0" id="quota-select-form" disabled="false" class="oc-mb-s" label="Personal quota" fix-message-line="true" description-message="" read-only="false"></quota-select-stub>
      <group-select-stub selectedgroups="" groupoptions="undefined,undefined" class="oc-mb-s" read-only="false"></group-select-stub>
    </div>
    <compare-save-dialog-stub originalobject="[object Object]" compareobject="[object Object]" confirmbuttondisabled="false" class="edit-compare-save-dialog oc-mb-l"></compare-save-dialog-stub>
  </form>
</div>"
`;
